src/main/kotlin/com/manimdsl/frontend/SymbolTableVisitor.kt

```diff
@@ -0,0 +1,42 @@
+package com.manimdsl.frontend
+
+data class IdentifierData(val type: Type)
+
+class SymbolTableVisitor {
+    private val scopes = mutableListOf<SymbolTable>(GlobalScopeSymbolTable())
+    private var currentScope: SymbolTable = scopes[0]
+
+    fun getTypeOf(identifier: String): Type {
+        return currentScope.table[identifier]?.type ?: (currentScope.parent?.table?.get(identifier)?.type ?: ErrorType)
```

Comment:
I'm still somewhat unsure this line is appropriate can we make design modification to make sure parent is not null

Reply:
I would disagree and say this is better? It is conventionally how a tree is defined (the root is a node with no parent) and it seems perfectly null safe (as there's no `!!`)

Reply:
When interoping with java nulls are inevitable but I think we should always prefer polymorphism when writing our own kotlin

Reply:
It could potentially be more than one line for readability sake? Do the null check first and then go ahead with a smart casted version?

Reply:
What was there before was a symboltableroot object which inherited from symbol table and it's "getTypeOf" just returns NoType so there's no checking and arguably much more readable

Reply:
We don't do the check it was basically the null pattern from SED before

Reply:
Like null safety isn't the concern it's more how the behaviour of root is defined by checking a null rather than basing it only on the  type.

Reply:
The nature of a hash map lookup is it returns the value or null so there's no real way round that, the symbol table has been cleaned up to simply be a hash map from string to identifier data so to do this you'd have to introduce unnecessary wrapper functions to the SymbolTable table variable along with creating a new Object that then always returns ErrorType

Reply:
The SED null pattern is one way round Null Pointer Exceptions for a language like Java, Kotlin's solution is the "?"

Reply:
Actually I don't think this would work as it does not recur through the parent's parent it just stops at the immediate parent

Reply:
We need a loop or to bring back the getTypeOf function in the SymbolTable

Reply:
You could use the id that's in the symbol table now?

Reply:
You could do something like this:

```
    fun getTypeOf(identifier: String, currentTable: SymbolTable = currentScope): Type {
        return currentTable.table[identifier]?.type ?: 
            if(currentTable.parent != null) getTypeOf(identifier, currentTable.parent) else ErrorType
    }
```

Reply:
This is good but I will have this be a private fun and expose a fun which does not take the currentTable.

Reply:
What do people think of [this](https://github.com/ManimDSL/ManimDSLCompiler/blob/feature/ch135/refactor-scoping-api-for-symbol-table/src/main/kotlin/com/manimdsl/frontend/SymbolTableVisitor.kt)?

Reply:
It has duplication but I think it is neater

Reply:
> The nature of a hash map lookup is it returns the value or null so there's no real way round that, the symbol table has been cleaned up to simply be a hash map from string to identifier data so to do this you'd have to introduce unnecessary wrapper functions to the SymbolTable table variable along with creating a new Object that then always returns ErrorType

Never a problem with the hash map lookup as I said before with interop with java it is appropriate

Reply:
That isn't what that pattern was. The pattern was not writing defensive code / branching code based on nulls but rather moving the functionality into the object and using polymorphism. In java I can equally do an if x == null like is done with ?

Reply:
I still find it confusing why there's a symbol table root and a global scope symbol table, isn't the global scope symbol table the top of the tree?

Reply:
`    fun getTypeOf(identifier: String, currentTable: SymbolTable = currentScope): Type {
        return currentTable.table[identifier]?.type ?:  getTypeOf(identifier, currentTable.parent)
    }`

This would become this

> Actually I don't think this would work as it does not recur through the parent's parent it just stops at the immediate parent

Where the null check happens it can just recurse into a parent object which returns back ErrorType. What do you mean by not recursing into parent's parent?

Reply:
> That isn't what that pattern was. The pattern was not writing defensive code / branching code based on nulls but rather moving the functionality into the object and using polymorphism. In java I can equally do an if x == null like is done with ?

The point is kotlin won't compile until that's handled, java you can, that's the whole reason they introduced the ? so that they can make sure you won't even run into null pointers

Reply:
The `SymbolTableRoot` is basically an error state. It means you have exited the global scope and is one level 'higher', so any variable lookup will return ErrorType

Reply:
If `GlobalScopeSymbolTable` is scope level 0, `SymbolTableRoot` is scope level -1

Reply:
Yup exactly @rbanerjee20

Reply:
> What do people think of [this](https://github.com/ManimDSL/ManimDSLCompiler/blob/feature/ch135/refactor-scoping-api-for-symbol-table/src/main/kotlin/com/manimdsl/frontend/SymbolTableVisitor.kt)?

You can go ahead with this if this solves the whole discussion and so it can be merged in, it's been in a state of almost merged for too long now

Reply:
I personally don't mind either way, whether there is polymorphism or not. However, the current implementation is wrong (our bad!), and I think with the implementation on (this branch)[https://github.com/ManimDSL/ManimDSLCompiler/blob/feature/ch135/refactor-scoping-api-for-symbol-table/src/main/kotlin/com/manimdsl/frontend/SymbolTableVisitor.kt], you have a recursive implementation which is more intuitive despite the duplication between `SymbolTableVisitor` and `SymbolTable` of `getTypeOf`.

What do people think? I agree that perhaps `SymbolTableRoot` is a bit of a misnomer since it actually represents an error scope.

Reply:
If we're reintroducint getTypeOf in the symbol table then we should introduce adding variables too because then accessing the table directly is "wrong" and can so table should be made private

Reply:
How is [this](https://github.com/ManimDSL/ManimDSLCompiler/blob/feature/ch135/refactor-scoping-api-for-symbol-table/src/main/kotlin/com/manimdsl/frontend/SymbolTableVisitor.kt)?

Reply:
It looks good to me!

Reply:
I think it looks good too! I just had some questions about if I can change the symbol table a bit for functions but we can merge this in first.

Reply:
Yep so the IdentifierData is open for you to extend for functions and add any other functions that are needed! 😄

Reply:
I think we're finally done. All we've changed from what I did on the refactor branch is instead of having a `SymbolTableRoot`, we treat `GlobalScopeSymbolTable` as the base case in the `getTypeOf` function - it searches for the variable in its own table and if not found, it returns `ErrorType`. So we finish the recursion at scope level 0 instead of having a scope level of -1 at all.

No null parent symbol tables and no duplication of `getTypeOf` and `addVariable` thanks to the overrided operator functions, which allow array-like getters and setters for symbol tables.

Please can someone approve if you are happy with what we have done!